README
Description
-----------

Search and display ISO codes for countries, languages, currencies, and scripts.

isoquery can be used to generate a tabular output of the ISO standard
codes provided by the package iso-codes. It parses the JSON files and shows
all included ISO codes or just matching entries, if specified on the command
line. Moreover, it's possible to get all available translations for
the ISO standard.

Dependencies
------------

In order to build the program from source, you need rst2man, po4a,
and gettext.

It's also very useful to have iso-codes installed, although this is
not strictly necessary.

Installation
------------

Just run "./configure && make install".

Author and Copyright
--------------------

Copyright © Dr. Tobias Quathamer <toddy@debian.org>

License
-------

This program is released under the GNU General Public License version 3,
see COPYING for details.


ChangeLog
Isoquery 3.2.3
==============
18 August 2018

* Fix FTBFS due to changed po4a string extraction in rst manpage
  Closes: #906475


Isoquery 3.2.2
==============
26 September 2017

* Fix FTBFS due to missing newline character in German translation.
  Closes: #876855


Isoquery 3.2.1
==============
30 August 2016

* Use Feature Test Macro for execvpe()


Isoquery 3.2.0
==============
30 August 2016

* Set up a defined environment to enable reproducible testing
* Update expected test result to match latest changes in iso-codes


Isoquery 3.1.0
==============
13 July 2016

* Updated translations:
  - Swedish, thanks to Martin Bagge / brother
  - Portuguese, thanks to Américo Monteiro
  - Vietnamese, thanks to Trần Ngọc Quân
  - Russian, thanks to Yuri Kozlov
  - Danish, thanks to Joe Hansen
  - French, thanks to Jean-Baka Domelevo Entfellner
* Updated manpage translations:
  - Portuguese, thanks to Américo Monteiro
  - French, thanks to Baptiste Jammet


Isoquery 3.0.1
==============
10 June 2016

* Fix uninitialized variable


Isoquery 3.0.0
==============
9 June 2016

* Complete rewrite in C, using GLib
* Supports the newly added ISO 3166-3 codes
* Updated translations:
  - German, thanks to Dr. Tobias Quathamer
* Updated manpage translations:
  - German, thanks to Dr. Tobias Quathamer


Isoquery 2.0
============
30 May 2014

* Complete rewrite in Vala, significantly faster than
  previous Python implementation
* Uses libisocodes to access iso-codes XML data
* Supports the newly added ISO 639-5 codes
* Updated translations:
  - Danish, thanks to Joe Hansen
  - Russian, thanks to Yuri Kozlov
  - French, thanks to Christian Perrier
  - Portuguese, thanks to Américo Monteiro
  - German, thanks to Tobias Quathamer
  - Swedish, thanks to Martin Bagge
* Updated manpage translations:
  - German, thanks to Tobias Quathamer
  - French, thanks to David Prévot
  - Portuguese, thanks to Américo Monteiro


Isoquery 1.7
============
23 January 2012

* Correct installation path for translated manpages.


Isoquery 1.6
============
8 January 2012

* Fix a UnicodeEncodeError in version output, thanks to David Prévot.
  Closes: #650746
* Use standard distutils setup.py file for building and installing
  the package, remove waf build system completely
* Updated translations:
  - Sinhala, thanks to Danishka Navin <danishka@gmail.com>.


Isoquery 1.5
============
14 June 2011

* If the locale does not exist for the given ISO standard, print a warning
  message and use the untranslated (thus English) strings. Closes: #624147
* Review isoquery manpage, thanks to David Prévot. Closes: #627577
* New manpage translation:
  - French, thanks to David Prévot. Closes: #628841
* Update test suite to pass with current iso-codes (3.26)
* Updated translations:
  - German, thanks to Tobias Quathamer.
  - Portuguese, thanks to Américo Monteiro.
  - Swedish, thanks to Martin Bagge. Closes: #628863
  - Russian, thanks to Yuri Kozlov. Closes: #628966
  - Czech, thanks to Michal Simunek. Closes: #629112
  - French, thanks to Christian Perrier. Closes: #629164
  - Danish, thanks to Joe Hansen. Closes: #630246


Isoquery 1.4
============
2 October 2010

* New translation:
  - Vietnamese, thanks to Clytie Siddall. Closes: #598618


Isoquery 1.3
============
15 August 2010

* New translation:
  - Portuguese, thanks to Américo Monteiro. Closes: #592431
* New manpage translation:
  - Portuguese, thanks to Américo Monteiro. Closes: #592433


Isoquery 1.2
============
31 July 2010

* Fix invalid syntax for Python 2.4. Thanks to Cristian Ionescu-Idbohrn
  for the bug report and patch. Closes: #591035


Isoquery 1.1
============
28 July 2010

* Fix errors in manpages with overlong lines
* Rewrite generation of manpages with waf build system
* French translation update, thanks to Christian Perrier. Closes: #589554
* New translations:
  - Russian, thanks to Yuri Kozlov. Closes: #589521
  - Swedish, thanks to Martin Bagge / brother <brother@bsnet.se>
  - Czech, thanks to Michal Šimůnek. Closes: #589719
  - Danish, thanks to Joe Hansen. Closes: #589861
  - Sinhala, thanks to Danishka Navin.
  - Spanish, thanks to Omar Campagne. Closes: #590564


Isoquery 1.0
============
30 June 2010

* Complete rewrite of program in Python
* Support of ISO 639-3 and ISO 3166-2


Isoquery 0.18
=============
1 June 2009

* Use gettext.h instead of libintl.h, as recommended by GNU gettext
* Update test suite to pass with current iso-codes (3.10)


Isoquery 0.17
=============
26 November 2008

* Change my last name to Quathamer
* Update test suite to pass with current iso-codes (3.4)


Isoquery 0.16
=============
1 July 2008

* Update test suite to pass with current iso-codes (3.1)


Isoquery 0.15
=============
2 June 2008

* Update man page with ISO 15924
* Fix bug that results in a crash if the environment variable
  LANGUAGE is not set
* Update test suite to pass with current iso-codes (3.0)


Isoquery 0.14
=============
24 April 2008

* Fix bug which could cause isoquery to throw an Glib::ConvertError.
  Thanks to George Danchev for the bug report. Closes: #474534


Isoquery 0.13
=============
12 March 2008

* Use C++ vectors for the list of possible xpaths
* New command line options "-0" and "--null" to separate the output
  with NULL characters instead of newline
* Translation updates:
  - French, thanks to Christian Perrier
  - German, thanks to Tobias Quathamer


Isoquery 0.12
=============
24 January 2008

* Add #include statements required by GCC 4.3. Thanks to
  Martin Michlmayr for the bug report and patch. Closes: #462206
* Better error message if the XML input file cannot be opened
* Translation updates:
  - French, thanks to Christian Perrier
  - German, thanks to Tobias Quathamer


Isoquery 0.11
=============
9 January 2008

* Require po4a in configure script for manpage translations
* Require pkg-config and libxml++ (>2.6.0) in configure script
* Add support for ISO 15924
* Update tests to pass with iso-codes 1.8

src/options.c
/**
 * Copyright © 2016 Dr. Tobias Quathamer <toddy@debian.org>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <stdlib.h>
#include <glib.h>
#include <glib/gi18n.h>
#include <glib/gprintf.h>
#include "options.h"

/**
 * Global variables to hold the program options.
 */
gchar *option_standard;
gchar *option_pathname;
gchar *option_namefield;
gchar *option_locale;
gboolean *option_null_separator;
gboolean *option_version;

// Helper variables, not to be accessed directly.
gboolean *option_name;
gboolean *option_officialname;
gboolean *option_commonname;

/**
 * Define the program command line options.
 */
static GOptionEntry entries[] = {
    {"iso", 'i', G_OPTION_FLAG_NONE, G_OPTION_ARG_STRING, &option_standard,
     N_
     ("The ISO standard to use. Possible values: 639-2, 639-3, 639-5, 3166-1, 3166-2, 3166-3, 4217, 15924 (default: 3166-1)"),
     N_("STANDARD")},
    {"pathname", 'p', G_OPTION_FLAG_NONE, G_OPTION_ARG_FILENAME, &option_pathname,
     N_("Use pathname as prefix for the data files (default: /usr/share/iso-codes/json)"),
     N_("PATHNAME")},
    {"locale", 'l', G_OPTION_FLAG_NONE, G_OPTION_ARG_STRING, &option_locale,
     N_("Use this locale for output"),
     N_("LOCALE")},
    {"name", 'n', G_OPTION_FLAG_NONE, G_OPTION_ARG_NONE, &option_name,
     N_("Name for the supplied codes (default)"), NULL},
    {"official_name", 'o', G_OPTION_FLAG_NONE, G_OPTION_ARG_NONE, &option_officialname,
     N_("Official name for the supplied codes. This may be the same as --name (only applies to ISO 3166-1)"),
     NULL},
    {"common_name", 'c', G_OPTION_FLAG_NONE, G_OPTION_ARG_NONE, &option_commonname,
     N_
     ("Common name for the supplied codes. This may be the same as --name (only applies to ISO 639-2, 639-3, and 3166-1)"),
     NULL},
    {"null", '0', G_OPTION_FLAG_NONE, G_OPTION_ARG_NONE, &option_null_separator,
     N_("Separate entries with a NULL character instead of newline"), NULL},
    {"version", 'v', G_OPTION_FLAG_NONE, G_OPTION_ARG_NONE, &option_version,
     N_("Show program version and copyright"), NULL},
    {NULL}
};

/**
 * Parse the command line arguments.
 *
 * @param arguments Array of arguments
 */
gboolean options_parse_command_line(gchar ** arguments, GError ** error)
{
    int argc = 0;
    GOptionContext *context;

    // Ensure that there are sensible default options.
    options_set_default_values();

    context = g_option_context_new(_("[ISO codes]"));
    g_option_context_add_main_entries(context, entries, GETTEXT_PACKAGE);

    // Count the number of arguments supplied.
    while (arguments[argc] != NULL) {
        argc++;
    }
    // If the parsing fails, return with error.
    if (!g_option_context_parse(context, &argc, &arguments, error)) {
        return FALSE;
    }
    // Validate options
    if (!options_validate(error)) {
        return FALSE;
    }
    return TRUE;
}

/**
 * Set default values for options.
 */
void options_set_default_values(void)
{
    option_standard = "3166-1";
    option_pathname = "/usr/share/iso-codes/json";
    option_namefield = "name";
    option_locale = "";
    option_null_separator = FALSE;
    option_version = FALSE;
}

/**
 * Validate values of options.
 */
gboolean options_validate(GError ** error)
{
    // Handle deprecated standards gracefully
    if (!g_strcmp0("3166", option_standard)) {
        option_standard = "3166-1";
        // Print a small warning
        g_printerr("isoquery: The standard 3166 is deprecated, please use 3166-1 instead.\n");
    }
    if (!g_strcmp0("639", option_standard)) {
        option_standard = "639-2";
        // Print a small warning
        g_printerr("isoquery: The standard 639 is deprecated, please use 639-2 instead.\n");
    }
    // Check that the given standard is supported
    int i = 0;
    gboolean supported = FALSE;
    gchar *supported_standards[] = { "639-2", "639-3", "639-5", "3166-1", "3166-2", "3166-3", "4217", "15924", NULL };
    while (supported_standards[i]) {
        if (!g_strcmp0(supported_standards[i], option_standard)) {
            supported = TRUE;
            break;
        }
        i++;
    }
    if (!supported) {
        // TRANSLATORS: The placeholder is a string like "1234-5".
        g_set_error(error, g_quark_from_string(GETTEXT_PACKAGE), 0, _("ISO standard \"%s\" is not supported."),
                    option_standard);
        return FALSE;
    }
    // Set the value for the name field
    if (option_officialname) {
        option_namefield = "official_name";
    }
    if (option_commonname) {
        option_namefield = "common_name";
    }
    return TRUE;
}

/**
 * Construct the filename for JSON data, given the pathname and ISO standard.
 *
 * @return gchar * Filename, has to be free'd by the caller.
 */
gchar *options_get_filename(void)
{
    gchar *filename, *complete_path;
    filename = g_strdup_printf("iso_%s.json", option_standard);
    complete_path = g_build_filename(option_pathname, filename, NULL);
    g_free(filename);
    return complete_path;
}

/**
 * Show program version and copyright
 */
void options_show_version(void)
{
    // TRANSLATORS: The placeholder is the version identifier.
    g_printf(_("isoquery %s\n"), VERSION);
    g_printf(_("Copyright © 2007-2016 Dr. Tobias Quathamer\n"));
    // TRANSLATORS: Please change the uppercase words as
    // appropriate for your language.
    gchar *translation = _("Translation to LANGUAGE Copyright © YEAR YOUR-NAME\n");
    if (!g_strrstr(translation, "LANGUAGE")) {
        g_printf("%s", translation);
    }
    g_print("\n");
    g_print("This program is free software: you can redistribute it and/or modify\n");
    g_print("it under the terms of the GNU General Public License as published by\n");
    g_print("the Free Software Foundation, either version 3 of the License, or\n");
    g_print("(at your option) any later version.\n");
    g_print("\n");
    g_print("This program is distributed in the hope that it will be useful,\n");
    g_print("but WITHOUT ANY WARRANTY; without even the implied warranty of\n");
    g_print("MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the\n");
    g_print("GNU General Public License for more details.\n");
    g_print("\n");
    g_print("You should have received a copy of the GNU General Public License\n");
    g_print("along with this program.  If not, see <http://www.gnu.org/licenses/>.\n");
}
